File: Medium/Python/Math/HappyNumber.py
class Solution(object):
    def isHappy(self, n):
        """
            :type n: int
            :rtype: bool
            """
        res = set()
        while n != 1:
            n = sum([int(num) ** 2 for num in str(n)])
            if n in res:
                return False
            res.add(n)
        return True

File: Medium/Java/Math/HappyNumber.java
class Solution {
    public boolean isHappy(int n) {
        Set<Integer> set = new HashSet<>();
        return happyPow(set, n) == 1? true: false;
    }
    
    public int happyPow(Set<Integer> set, int n) {
        int res = 0;
        while (n != 0) {
            res += Math.pow(n % 10, 2);
            n /= 10;
        }
        if (!set.contains(res)) {
            set.add(res);
            res = happyPow(set, res);
        }
        return res;
    }
}
